### PR TITLE
[rh:niquests] Add support for the `niquests` HTTP library

### PR DESCRIPTION
**Pull request by Ousret**
### Description of your *pull request* and other information

Add support for Niquests as an optional network handler.
Supersede PR https://github.com/yt-dlp/yt-dlp/pull/8624

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [x] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by Ousret** [comment]
Everything should be in order and ready for merge.
I am waiting upon your review.

**Comment by Ousret** [comment]
gentle ping. is there anything that can be done on my side to ease with the review?

for a side note, HTTP/3 support with Niquests was greatly improved. And I can say with confidence that we are able to mimic Firefox HTTP3 TLS fingerprint and we can adapt in case of changes.

regards

**Review by Grub4K — COMMENTED** [review]
Just some initial looking over it. You have to be patient for @coletdjnz to do an actual review.

Also, while the implementations for both `niquests` and `requests` look strikingly similar, many of the similarities are broken since `import niquests_urllib3 as urllib3` gets done at the top of the file. Im not sure there are very many things that could be factored out?

**Comment by coletdjnz** [comment]
i have this planned to look at after I finish ironing out a couple other minor issues. Please be aware things tend to move slowly here when touching core code (from my experience).

**Comment by Ousret** [comment]
Noted. I was mistaken when I though that having proper guidance on a optional feature wouldn't be too difficult to merge. Especially having the test scenarios copied from "requests", which you already trust.
I'll wait for a complete review before moving forward.

Regards,

File: yt_dlp/networking/_niquests.py
Comment (Ousret) [reply yt_dlp/networking/_niquests.py:94]: OK. Will do.

File: devscripts/install_deps.py
Comment (Ousret) [reply devscripts/install_deps.py:81]: Noted.

File: devscripts/install_deps.py
Comment (Ousret) [reply devscripts/install_deps.py:78]: Does this is a style preference? I find it less readable.

File: yt_dlp/YoutubeDL.py
Comment (coletdjnz) [inline yt_dlp/YoutubeDL.py:4155]:
```suggestion
                    and 'Niquests' not in self._request_director.handlers
```

(this is a bug I'm fixing in another PR, but just to prepare for the patch...)

**Review by coletdjnz — CHANGES_REQUESTED** [review]
One issue we have become aware of:

https://niquests.readthedocs.io/en/latest/community/faq.html#what-is-urllib3-futur

> Your environment may or may not include the legacy urllib3 package in addition to urllib3.future. So doing:
> import urllib3
> May actually import either urllib3 or urllib3.future. But fear not, if your script was compatible with urllib3, it will most certainly work out-of-the-box with urllib3.future.
> This behavior was chosen to ensure the highest level of compatibility for your migration, ensuring the minimum friction during the migration between Requests to Niquests.

Our use case requires Requests to co-exist with Niquests, with urllib3-future not overriding urllib3. Our request handlers must not interfere with each other (able to co-exist) and need to be able to be tested independently (I am not able to accurately test this code and we cannot trust the requests tests when urllib3.future is installed, which is vital for the CI). We also have additional compatibility requirements when it comes to embedding etc.

Therefore we would need urllib3-future (or another package) to have it's own namespace and not override urllib3. Installing them in a particular order is not a solution and will not be feasible to enforce in most situations.

https://github.com/jawah/urllib3.future/issues/91

Unfortunately this is probably going to be a blocker for us to accept this (and probably with many other projects too you find), so I encourage you to reconsider this.

Perhaps consider the PyCryptodome and PyCryptodomeX model, where the former is like urllib3-future and injects itself into the Crypto package, whereas the latter doesn't and has it's own independent namespace. You could maybe provide the equivalent of both?
https://www.pycryptodome.org/src/introduction#pycryptodome

File: yt_dlp/networking/_niquests.py
Comment (Ousret) [reply yt_dlp/networking/_niquests.py:273]: Noted for the style preference. My mistake.

> Also don't think it is required to pass this header to niquests? Since it uses http/2 and the like?

No, it's actually illegal header for http 2+. it is quietly removed deep in the stack.

File: yt_dlp/networking/_niquests.py
Comment (Ousret) [reply yt_dlp/networking/_niquests.py:337]: Yes.

File: yt_dlp/YoutubeDL.py
Comment (Ousret) [reply yt_dlp/YoutubeDL.py:4155]: Noted.

File: yt_dlp/networking/_niquests.py
Comment (Ousret) [reply yt_dlp/networking/_niquests.py:167]: I was uncomfortable doing the breakage on this one. Found too much project relying on broken behaviors. Still considering it though.

**Comment by Ousret** [comment]
Looking at your main comment.

> Therefore we would need urllib3-future (or another package) to have it's own namespace and not override urllib3. Installing them in a particular order is not a solution and will not be feasible to enforce in most situations.

https://github.com/jawah/urllib3.future/issues/91

> Unfortunately this is probably going to be a blocker for us to accept this (and probably with many other projects too you find), so I encourage you to reconsider this.

I am not going to change my mind on this one. If you want to make sure the legacy urllib3 is installed (in CI), there are many ways that can be explored.

End-users aren't likely to want both (http core) engine in their package tree. I've taken many long weeks thinking of this, and the explanations given on both linked issue and linked documentation are enough to explain the use cases. I am confident that this is the right choice even if its not perfect. There is a plan to change this, but in a distant future.

As you did not plan to include Niquests by default, it should not be a problem. But again,

You may close this one if we agree to disagree on the "blocker". If you're keeping the door open, I will promptly fix the remaining comments.

regards,

**Comment by Grub4K** [comment]
Providing two packages, where one does not replace is the superior way in all circumstances. Your reasoning to unblock old packages is covered by this (they can continue to install `urllib3.future`), as well as having reproducible and separate dependencies (installing `urllib3.xfuture` or however that package would be named). I would also argue this is not much work and I can provide a PR for this workflow on your side if wanted.

If you decide to *still* think that providing two packages does not make sense, then we unfortunately cannot move this forward. I hope you reconsider

**Comment by Ousret** [comment]
> Providing two packages, where one does not replace is the superior way in all circumstances.

Your opinion. You haven't seen the big picture, yet. 

> I would also argue this is not much work and I can provide a PR for this workflow on your side if wanted.

It's disconcerting, for seven months, I was asked to do efforts in order to do a high effort to lower the maintenance burden on you. Now I am asked to make my burden considerably higher. Me too, can provide a PR where I replace Requests without too much work. _spoiler: almost none_ but you would decline on...? maintenance burden?

> If you decide to still think that providing two packages does not make sense, then we unfortunately cannot move this forward. I hope you reconsider

As far as I can see, it is a breaking change and/or too high burden when I am almost on my limits.

yt-dlp already have requests and curl via cffi, it is more than enough.

thanks for considering it though.

regards,

**Comment by Grub4K** [comment]
It's not a breaking change, and you haven't explained the big picture anywhere.

We can reconsider once the library has matured enough. Thank you for your efforts!

**Pull request closed without merging**
